### PR TITLE
Add json rpc starknet_traceTransaction

### PR DESCRIPTION
**Pull request by mikiw**
## Usage related changes

- Add of starknet_traceTransaction json rpc endpoint

## Development related changes

- refactor of `try_from_call_info()`

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] No unused dependencies - `./scripts/check_unused_deps.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Checked the TODO section in README.md if this PR resolves it
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

**Comment by mikiw** [comment]
TODO on Monday: Add declare, deploy + tests. Refactor common code that it's also in simulate function.

File: crates/starknet-server/tests/test_dump_and_load.rs [hunk 1]
```diff
@@ -1,6 +1,5 @@
 pub mod common;
 
-// Important! Use unique file names for dump files, tests can be run in parallel.
```
Comment (mikiw) [inline crates/starknet-server/tests/test_dump_and_load.rs:3]: this comment is not needed anymore

File: crates/types/src/rpc/transactions.rs [hunk 2]
```diff
@@ -435,11 +435,11 @@ pub struct SimulatedTransaction {
 
 impl FunctionInvocation {
     pub fn try_from_call_info(
-        mut call_info: blockifier::execution::call_info::CallInfo,
+        call_info: &blockifier::execution::call_info::CallInfo,
```
Comment (FabijanC) [inline crates/types/src/rpc/transactions.rs:438]: Why was this needed?

File: crates/starknet/src/starknet/mod.rs [hunk 3]
```diff
@@ -705,6 +706,107 @@ impl Starknet {
...
+                    }
+                }
+            }
+        })
+    }
+
+    fn get_call_info_invocation(
+        call_info_invocation: &Option<CallInfo>,
+        address_to_class_hash_map: &HashMap<ContractAddress, ClassHash>,
+    ) -> DevnetResult<Option<FunctionInvocation>> {
```
Comment (FabijanC) [inline crates/starknet/src/starknet/mod.rs:746]: Wrapping an `Option` in a `(Devnet)Result` looks wrong. Also, this implementation looks like it could benefit from methods of `Option`.

File: crates/starknet/src/starknet/mod.rs [hunk 4]
```diff
@@ -705,6 +706,107 @@ impl Starknet {
...
+    ) -> DevnetResult<TransactionTrace> {
+        let transaction_to_map =
+            self.transactions.get(&transaction_hash).ok_or(Error::NoTransaction)?;
+
+        let mut state =
+            self.get_state_at(&BlockId::Tag(starknet_rs_core::types::BlockTag::Latest))?.clone();
+
+        let state_diff: ThinStateDiff = state.extract_state_diff_from_pending_state()?.into();
+        let state_diff =
+            if state_diff == ThinStateDiff::default() { None } else { Some(state_diff) };
```
Comment (FabijanC) [inline crates/starknet/src/starknet/mod.rs:766]: This is already the second time that we are introducing this hacky check.
It's been discussed in [this conversation](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/257/files#r1401714563) with @marioiordanov.

Maybe more logic could be extracted to a common function used by `get_transaction_trace` and `simulate_transactions`?

**Review by FabijanC — CHANGES_REQUESTED** [review]

**Review by marioiordanov — CHANGES_REQUESTED** [review]
In starknet-server/test_data/spec/0.5.1/starknet_trace_api_openrpc.json the method starknet_traceTransaction is missing, because until this PR there was no need for that. Please add it and check if serialization/deserialization works for it, by running `spec_reader/mod.rs` tests

File: crates/types/src/rpc/transactions.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/types/src/rpc/transactions.rs:438]: `starknet_traceTransaction` is invoked for already executed transactions. They are stored in collection and getting them from it returns a reference. The reason for changing the signature of the method is that `blockifier::execution::call_info::CallInfo` doesnt implement Copy/Clone and using it with an object that is a reference struct results in `moving out of shared reference` error.

File: crates/types/src/rpc/transactions.rs [hunk 5]
```diff
@@ -450,23 +450,28 @@ impl FunctionInvocation {
...
-
-        let messages: Vec<OrderedMessageToL1> = call_info
+        let mut messages: Vec<OrderedMessageToL1> = call_info
             .execution
             .l2_to_l1_messages
-            .into_iter()
-            .map(|msg| OrderedMessageToL1::new(msg, call_info.call.caller_address.into()))
+            .iter()
+            .map(|msg| {
+                OrderedMessageToL1::new(
```
Comment (marioiordanov) [inline crates/types/src/rpc/transactions.rs:458]: `OrderedMessageToL1::new` method is used only in this place. Why not refactoring it to accept `&OrderedL2ToL1Message` instead of `OrderedL2ToL1Message`.

File: crates/starknet/src/starknet/mod.rs (hunk 3 above)
Comment (mikiw) [reply crates/starknet/src/starknet/mod.rs:746]: @marioiordanov? I was just refactoring your code :)

Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs:746]: It might just be my Rust inexperience and this pattern is considered normal, but it does look suspicious.

File: crates/starknet/src/starknet/mod.rs (hunk 4 above)
Comment (mikiw) [reply crates/starknet/src/starknet/mod.rs]: fixed in 2fac26806808f07a966160c7249bb15458169ce3

Comment (mikiw) [reply crates/starknet/src/starknet/mod.rs:766]: Ok not fixed and will not be fixed I guess...

I needed to revert that in f888865b0350b62903ce7d246c2437c806b39cee. The problem is that this refactor will cause extra clones of that state :(

Additionally, it gave me a headache (×_×;）

**Comment by mikiw** [comment]
> In starknet-server/test_data/spec/0.5.1/starknet_trace_api_openrpc.json the method starknet_traceTransaction is missing, because until this PR there was no need for that. Please add it and check if serialization/deserialization works for it, by running `spec_reader/mod.rs` tests

Added in b30756459ad5516382be568bbf2e866c2f13e5dd

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 6]
```diff
@@ -374,4 +374,17 @@ impl JsonRpcHandler {
...
+    /// starknet_traceTransaction
+    pub(crate) async fn get_trace_transaction(
+        &self,
+        transaction_hash: TransactionHash,
+    ) -> StrictRpcResult {
+        let starknet = self.api.starknet.read().await;
+        match starknet.get_transaction_trace_by_hash(transaction_hash) {
+            Ok(result) => Ok(StarknetResponse::TraceTransaction(result)),
+            Err(Error::NoTransaction) => Err(ApiError::TransactionNotFound),
+            Err(err) => Err(err.into()),
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: So I added `NO_TRACE_AVAILABLE` but I'm not sure why are still not handling `INVALID_TXN_HASH`?

File: crates/starknet/src/starknet/mod.rs (hunk 3 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs:746]: > @marioiordanov? I was just refactoring your code :)

What should be my expected reply

Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs:746]: > What should be my expected reply

One of these:
- telling us of a magical way to refactor this not to use Results wrapped in Options
- telling us that this is a common pattern and we lack experience

Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs:746]: Ahh, okay. ```rust
get_call_info_invocation(call_info_invocation: &CallInfo,
        address_to_class_hash_map: &HashMap<ContractAddress, ClassHash>)-> DevnetResult<FunctionInvocation>``` . By removing the option from call_info_invocation it is removed from the result.
        
 Before executing get_call_info_invocation, the match statement will be in the form:
 ```rust
if let Some(ref x) = transaction_to_map.execution_info.validate_call_info {
            Some(FunctionInvocation::try_from_call_info(x, address_to_class_hash_map)?)
        } else {
            None
        };
 ```

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: ok for now I added something like this 3b2692f0ebf0dde4c7b852f056c372ef50e408ad is that ok?

[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: I would leave it as it is for now, once there will be reject support or better specification we can change it again

[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: Ok, since we are never rejecting, and our txs are never RECEIVED, we can resolve this.

**Review by FabijanC — APPROVED** [review]
Giving my approval, but let's resolve the remaining refactoring request

**Comment by mikiw** [comment]
L1hanlder transactions support will be added in https://github.com/0xSpaceShard/starknet-devnet-rs/issues/280

File: crates/starknet/src/starknet/mod.rs (hunk 3 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs:760]: For me it doesn't look good to be Result<Option<T>> but If there is no other way I will resolve this conversation. Also creating an issue for this refactoring doesn't make sense, because this part was refactored already in this PR.

**Review by marioiordanov — APPROVED** [review]
Approving, but lets refer to this comment https://github.com/0xSpaceShard/starknet-devnet-rs/pull/266/files#r1420574368
